Libraries/src/net/relinc/libraries/data/DataSubset.java

```diff
@@ -2,55 +2,65 @@
 
 import net.relinc.libraries.data.ModifierFolder.Modifier;
 import net.relinc.libraries.data.ModifierFolder.ModifierListWrapper;
+import net.relinc.libraries.data.ModifierFolder.ModifierResult;
+import net.relinc.libraries.data.ModifierFolder.Resampler;
 import net.relinc.libraries.staticClasses.SPSettings;
 
+import java.util.Optional;
+
 public abstract class DataSubset {
-	private int begin;
+	// I need to keep track of the original begin/end so I can save them to disk... then just use getBegin() and getEnd() to scale to sampled data
+	private int begin; // always refers to the original data... then they need to get scaled as necessary..
 	private int end;
 	private Integer beginTemp; //set in viewer
 	private Integer endTemp; //set in viewer
+
+	private Optional<ModifierResult> modifierResult = Optional.empty();
+
 	public String name = "";
-	public Dataset Data;
+	private Dataset Data;
 	public DataFileInfo fileInfo;
-	public ModifierListWrapper modifiers;
+	private ModifierListWrapper modifiers;
+	// This gets defaulted to the original amount of data points. The user may choose to make it higher or lower, in which case it is interpolated.
 	
 	public abstract baseDataType getBaseDataType();
 	
 	public abstract String getUnitAbbreviation(); // Each DataSubset has its standard units. e.g. Force=N
 	public abstract String getUnitName(); // e.g. Force = Newtons
-	
-	public void reduceDataNonReversible(int pointsToKeep) {
-		// This is non-reversible bowling-ball that modifies all the data... 
-		int reductionFactor = this.Data.timeData.length / pointsToKeep;
-		if(reductionFactor <= 1)
-		{
-			// TODO: log something somewhere...
-			return;
+
+	public ModifierResult getModifierResult() {
+
+		if(modifierResult.isPresent()) {
+			return modifierResult.get();
 		}
-		
-		double[] newTimeData = new double[pointsToKeep];
-		double[] newData = new double[pointsToKeep];
-		for(int i = 0; i < pointsToKeep; i++)
-		{
-			newTimeData[i] = this.Data.timeData[i * reductionFactor];
-			newData[i] = this.Data.data[i * reductionFactor];
+		this.render();
+		if(modifierResult.isPresent()) {
+			return modifierResult.get();
+		} else {
+			throw new RuntimeException("Expected modifier result to be rendered!");
 		}
-		
-		this.Data.timeData = newTimeData;
-		this.Data.data = newData;
-		
-		this.setBegin(this.begin / reductionFactor);
-		this.setEnd(this.end / reductionFactor);
-		if(this.getBeginTemp() != null)
-			this.setBeginTemp(this.getBeginTemp() / reductionFactor);
-		if(this.getEndTemp() != null)
-			this.setEndTemp(this.getEndTemp() / reductionFactor);
-		
+	}
+
+	public void invalidateResult() {
+		modifierResult = Optional.empty();
+	}
+
+	public ModifierListWrapper getModifiers() {
+		// they might change a modifier, so we have to invalidate the result.
+		this.modifierResult = Optional.empty();
```

Comment:
if someone is getting the modifiers, they could be altering them.. this is really just a leaky way to try to invalidate the cache. It doesn't work in all cases, if they keep a reference to the list of modifiers around, it doesn't invalidate the cache properly